lib/rack/builder.rb

```diff
@@ -93,28 +60,50 @@ def self.load_file(path)
 
       config.sub!(/^__END__\n.*\Z/m, '')
 
-      return new_from_string(config, path)
+      return new_from_string(config, path, **options)
     end
 
     # Evaluate the given +builder_script+ string in the context of
     # a Rack::Builder block, returning a Rack application.
-    def self.new_from_string(builder_script, file = "(rackup)")
-      # We want to build a variant of TOPLEVEL_BINDING with self as a Rack::Builder instance.
-      # We cannot use instance_eval(String) as that would resolve constants differently.
-      binding, builder = TOPLEVEL_BINDING.eval('Rack::Builder.new.instance_eval { [binding, self] }')
-      eval builder_script, binding, file
+    def self.new_from_string(builder_script, file = "(rackup)", **options)
+      builder = self.new(**options)
+
+      # Create a top level scope with self as the builder instance:
+      binding = TOPLEVEL_BINDING.eval('->(builder){builder.instance_eval{binding}}').call(builder)
+      
+      eval(builder_script, binding, file)
 
       return builder.to_app
     end
 
     # Initialize a new Rack::Builder instance.  +default_app+ specifies the
     # default application if +run+ is not called later.  If a block
     # is given, it is evaluted in the context of the instance.
-    def initialize(default_app = nil, &block)
-      @use, @map, @run, @warmup, @freeze_app = [], nil, default_app, nil, false
+    def initialize(default_app = nil, **options, &block)
+      @use = []
+      @map = nil
+      @run = default_app
+      @warmup = nil
+      @freeze_app = false
+
+      @options = options
+
       instance_eval(&block) if block_given?
     end
 
+    attr :options
+
+    # Whether the application server will invoke the application from multiple threads. Implies {reentrant?}.
```

Comment:
`multithread?` and `reentrant?` are not strictly related to the removal of `rack.multithread`/`rack.multiprocess`/`rack.run_once` and I also think they should not be part of the same pull request.

Reply:
They are related in the sense that we need to replace with something else but sure it can be a separate PR. I'll split it up.